File: apps/processor/test/embedding.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { embedText, fetchEmbedding } from "../src/embedding.js";

test("builds normalized local embeddings", () => {
	const vector = embedText("Astra keeps field notes");

	assert.equal(vector.length, 384);
	const magnitude = Math.sqrt(vector.reduce((sum, value) => sum + value * value, 0));
	assert.ok(Math.abs(magnitude - 1) < 0.000001);
});

test("uses local embeddings when no endpoint is configured", async () => {
	const embedding = await fetchEmbedding("", "local-embeddings", "Astra keeps field notes");

	assert.equal(embedding.model, "local-embeddings");
	assert.equal(embedding.vector.length, 384);
});

test("requires an endpoint for non-local models", async () => {
	await assert.rejects(
		fetchEmbedding("", "remote-model", "Astra keeps field notes"),
		/EMBEDDING_ENDPOINT is required/,
	);
});

File: apps/pkg/embeddingutil/embedding_test.go
package embeddingutil

import (
	"math"
	"testing"
)

func TestEmbedTextReturnsNormalizedVector(t *testing.T) {
	vector := EmbedText("Astra keeps field notes")
	if len(vector) != Dimensions {
		t.Fatalf("expected %d dimensions, got %d", Dimensions, len(vector))
	}

	var sum float64
	for _, value := range vector {
		sum += value * value
	}
	if math.Abs(math.Sqrt(sum)-1) > 0.000001 {
		t.Fatalf("expected normalized vector, got magnitude %f", math.Sqrt(sum))
	}
}

func TestEmbedTextIsDeterministic(t *testing.T) {
	first := EmbedText("Astra keeps field notes")
	second := EmbedText("Astra keeps field notes")

	for index := range first {
		if first[index] != second[index] {
			t.Fatalf("expected deterministic value at index %d", index)
		}
	}
}
